### ISSUE TITLE
Creation of HKLM\Software\Wow6432Node\Wine registry key hides DllOverrides for 32-bit applications (Wow64 prefix, triggered by GTA IV)

### ISSUE THREAD

**Issue by karolherbst** [wine-compholio/wine-staging#270]
I had to recreate my steam prefix twice already. After 

1. winetricks steam
2. I've only copied over my old steamapps folder
3. removed wine-mono
4. installed native .NET 4.0
5. played "DmC: Devil may Cry" (which needs native .NET 4.0)
6. played around with GTA IV graphic settings.

since then the user.reg file doesn't seem to get loaded, effectivly ignoring all DLLOverides. It could be a different issue though, but I don'T see the DLLOverrides section inside regedit and a lot of user settings are gone (like some GTA IV serial key stuff and so on).

I didn't see anything usefull from the normal console output.

**Comment by slackner** [comment]
Hi @karolherbst,

Nothing is really special about the commands you executed. I assume you know that it is recommended to install and run each application into a different wine prefix (to avoid conflicts of different runtime dlls), but its unlikely that this causes problems here.

I have never heard of this so problem so far, but not sure if it is really a bug in wine. If you for example forcefully terminated the `wineserver`, there is a specific chance of corrupting the registry files (both with regular wine and wine-staging), when you do that at the wrong time. It would also be possible that you were probably using multiple wineprefixes, and used the wrong one for starting Steam. Wine will always use the settings of the wineprefix specified by the `WINEPREFIX` environment variable, not the directory where the applications are in.

Why do you think that DllOverrides is ignored? How do you test that, and which overrides do you have added, which are not respected? Try to create a log for example with `WINEDEBUG=+loaddll`, this might be useful to debug overrides and stuff.

Regards,
Sebastian

**Comment by karolherbst** [comment]
It happend now the third time through GTA, so I guess it is unlikly that it has something to do with anything else.

Also WINEDLLOVERRIDES works if I set it. I've noticed it, because steam began to load dwrite again, which is really obvious, because Steam don't have any fonts then and dwrite is disabled in winecfg, which is also a weird sympton, because winecfg seems to parse the overrides, but everything else does not.

Also I have a script to start steam inside wine, where all needed variables are exported, also my mscoree override (for native .NET) gets ignored.

**Comment by karolherbst** [comment]
here is the output with WINEDEBUG=+loaddll,-dwrite

dwrite just throws out too many lines, so I disabled it.

https://gist.github.com/karolherbst/00f278873c9d3b5eeb85

**Comment by karolherbst** [comment]
this is part of my user.reg

[Software\\Wine\\DllOverrides] 1420806703
"*dwrite"="\0"
"*gameoverlayrenderer"="\0"
"*mscoree"="native\0"
"xinput1_3"="native"

**Comment by karolherbst** [comment]
but regedt doesn't show it anywhere

**Comment by slackner** [comment]
Did you double-check that `WINEPREFIX` is set properly, and that the instance of steam you are running is a subdirectory of the prefix?

The log doesn't contain anything really useful so far, please try to create one with:
`WINEDEBUG=+loaddll,+module,-dwrite`

This will enable a couple of other messages which probably show whats going wrong.

**Comment by karolherbst** [comment]
"Did you double-check that WINEPREFIX is set properly, and that the instance of steam you are running is a subdirectory of the prefix?" yes.

here is the new output: 
https://gist.github.com/karolherbst/0e8ce4d36af399c55349
be aware, that there are two files, where the second one isn't shown because of character limit.

**Comment by slackner** [comment]
```
trace:module:get_load_order looking up loadorder for L"C:\\windows\\system32\\mscoree.dll"
trace:module:get_load_order_value got standard key n for L"*mscoree"
```

For some of your overrides it is recognized properly, but really weird why the dwrite override doesn't have any effect. Will do some testing myself and then report back.

**Comment by slackner** [comment]
I did a couple tests with dlloverrides in a fresh wineprefix, and everything seems to work exactly as expected so far. I think we definitely need more information to be able to help in this case.

* Which wine version and distribution are you using? Based on your log its not a pure wine-staging build, if you have also other patches applied those could also be responsible for your issue. See:
```
Native Direct3D 9 is disabled.
For more information visit https://wiki.ixit.cz/d3d9
```
 
* Could you please provide the exact instructions you use for creating the wineprefix / running steam and so on? Then we can try to reproduce it step-by-step.

* Did you try with an unpatched version of Wine? This is probably the fastest way to find out if its a regression by our patchset, or something else.

**Comment by karolherbst** [comment]
I am using wine-staging-1.7.34 with these patches applies: https://github.com/karolherbst/F.U.N.-overlay/tree/master/app-emulation/wine/files

the Xinput patches are for xinput1_3 support, the wine-1.7.33-d3d9adapter.patch is for the native d3d support through the mesa gallium nine state tracker (for testing), wine-ToUnicodeEx-dead-key.patch is for a bug in "divinity: original sin" (see: https://bugs.winehq.org/show_bug.cgi?id=36854) all the others are patches applied on normal gentoo packages. But I doubt any of them would trigger this bug.

"Could you please provide the exact instructions you use for creating the wineprefix / running steam and so on? Then we can try to reproduce it step-by-step."
Basically it happens after toying around with the graphical settings inside GTA IV. Everything else would be highly unlikely, because I never had the issue before and was using the prefix a long time. And it happens three times after toying around with the settings there.

"Did you try with an unpatched version of Wine? This is probably the fastest way to find out if its a regression by our patchset, or something else." Not yet, I will do so now.

**Comment by karolherbst** [comment]
the wine-1.7.28-gstreamer-v4.patch wasn't use by me

**Comment by titan73** [comment]
You can delete this ones as they are empty and thus not taken into account
```
"*dwrite"="\0"
"*gameoverlayrenderer"="\0"
```

For mscoree:
```
"*mscoree"="native\0"
trace:module:get_redirect looking up redirection for L"C:\\windows\\system32\\mscoree.dll"
trace:module:get_redirect no redirection found for L"C:\\windows\\system32\\mscoree.dll"
trace:module:get_load_order looking up loadorder for L"C:\\windows\\system32\\mscoree.dll"
trace:module:get_load_order_value got standard key n for L"*mscoree"
```
Wine sees the override,
```
trace:module:load_dll L"C:\\windows\\system32\\mscoree.dll" is a fake Wine dll
```
tries to load the native but the native dll does not exist
```
warn:module:load_dll Failed to load module L"mscoree.dll"; status=c0000135
```
so it disables this dll. You need to install  the native dll

For xinput1_3:
```
"xinput1_3"="native"
trace:module:find_ordinal_export ordinal 100 out of range!
trace:module:find_forwarded_export delay loading L"xinput1_3.dll" for 'xinput1_3.XInputSetState'
trace:module:load_dll looking for L"xinput1_3.dll" in L"C:\\Program Files (x86)\\Steam;.;C:\\windows\\system32;C:\\windows\\system;C:\\windows;C:\\windows\\system32;C:\\windows;C:\\windows\\system32\\wbem"
trace:module:get_redirect looking up redirection for L"C:\\windows\\system32\\xinput1_3.dll"
trace:module:get_redirect no redirection found for L"C:\\windows\\system32\\xinput1_3.dll"
trace:module:get_load_order looking up loadorder for L"C:\\windows\\system32\\xinput1_3.dll"
trace:module:get_load_order got hardcoded default for L"C:\\windows\\system32\\xinput1_3.dll"
```
Wine does not find the override ("get_load_order_value got standard key n" missing)
why is the ending \0 missing compared to the others? Does it help adding it?
```
trace:module:load_builtin_dll Trying built-in L"C:\\windows\\system32\\xinput1_3.dll"
warn:module:load_builtin_dll failed to load .so lib for builtin L"C:\\windows\\system32\\xinput1_3.dll": /home/karol/.local/share/wineprefixes/steam/dosdevices/c:/windows/syswow64/xinput1_3.dll: invalid ELF header
```
This line is a bit weird but it does not seem to matter
```
trace:module:load_builtin_dll Trying built-in L"xinput1_3.dll"
trace:module:load_dll looking for L"kernel32.dll" in L"C:\\Program Files (x86)\\Steam;.;C:\\windows\\system32;C:\\windows\\system;C:\\windows;C:\\windows\\system32;C:\\windows;C:\\windows\\system32\\wbem"
trace:module:load_dll Found L"C:\\windows\\system32\\KERNEL32.dll" for L"kernel32.dll" at 0x7b810000, count=-1
trace:module:load_builtin_callback loaded xinput1_3.dll 0x32a19d8 0x7bf90000
trace:loaddll:load_builtin_dll Loaded L"C:\\windows\\system32\\xinput1_3.dll" at 0x7bf90000: builtin
trace:module:MODULE_InitDLL (0x7bf90000 L"xinput1_3.dll",WINE_PREATTACH,(nil)) - CALL
trace:module:MODULE_InitDLL (0x7bf90000,WINE_PREATTACH,(nil)) - RETURN 0
```
The builtin dll is loaded and tells wine that if there is a native version it should be used instead
```
trace:module:load_dll L"C:\\windows\\system32\\xinput1_3.dll" pre-attach returned FALSE, preferring native
trace:module:LdrUnloadDll (0x7bf90000)
trace:module:LdrUnloadDll (L"xinput1_3.dll") - START
trace:module:MODULE_DecRefCount (L"xinput1_3.dll") ldr.LoadCount: 0
trace:module:free_modref unloading L"C:\\windows\\system32\\xinput1_3.dll"
trace:module:LdrUnloadDll END
trace:module:load_native_dll Trying native dll L"C:\\windows\\system32\\xinput1_3.dll"
trace:module:map_image mapped PE file at 0xe0a0000-0xe0b6000
trace:module:map_image mapping section .text at 0xe0a1000 off 400 size f000 virt ef6a flags 60000020
trace:module:map_image mapping section .data at 0xe0b0000 off f400 size c00 virt 2b1c flags c0000040
trace:module:map_image clearing 0xe0b0c00 - 0xe0b1000
trace:module:map_image mapping section .rsrc at 0xe0b3000 off 10000 size 400 virt 3c0 flags 40000040
trace:module:map_image clearing 0xe0b3400 - 0xe0b4000
trace:module:map_image mapping section .reloc at 0xe0b4000 off 10400 size 1600 virt 1516 flags 42000040
trace:module:map_image clearing 0xe0b5600 - 0xe0b6000
trace:module:map_image relocating from 0x400000-0x416000 to 0xe0a0000-0xe0b6000
warn:module:alloc_module disabling no-exec because of L"xinput1_3.dll"
trace:module:load_dll looking for L"ntdll.dll" in L"C:\\Program Files (x86)\\Steam;.;C:\\windows\\system32;C:\\windows\\system;C:\\windows;C:\\windows\\system32;C:\\windows;C:\\windows\\system32\\wbem"
trace:module:load_dll Found L"C:\\windows\\system32\\ntdll.dll" for L"ntdll.dll" at 0x7bc10000, count=-1
trace:module:load_dll looking for L"KERNEL32.dll" in L"C:\\Program Files (x86)\\Steam;.;C:\\windows\\system32;C:\\windows\\system;C:\\windows;C:\\windows\\system32;C:\\windows;C:\\windows\\system32\\wbem"
trace:module:load_dll Found L"C:\\windows\\system32\\KERNEL32.dll" for L"KERNEL32.dll" at 0x7b810000, count=-1
trace:module:load_dll looking for L"ADVAPI32.dll" in L"C:\\Program Files (x86)\\Steam;.;C:\\windows\\system32;C:\\windows\\system;C:\\windows;C:\\windows\\system32;C:\\windows;C:\\windows\\system32\\wbem"
trace:module:load_dll Found L"C:\\windows\\system32\\advapi32.dll" for L"ADVAPI32.dll" at 0x7e990000, count=-1
trace:module:load_dll looking for L"SETUPAPI.dll" in L"C:\\Program Files (x86)\\Steam;.;C:\\windows\\system32;C:\\windows\\system;C:\\windows;C:\\windows\\system32;C:\\windows;C:\\windows\\system32\\wbem"
trace:module:load_dll Found L"C:\\windows\\system32\\setupapi.dll" for L"SETUPAPI.dll" at 0x7d230000, count=2
trace:loaddll:load_native_dll Loaded L"C:\\windows\\system32\\xinput1_3.dll" at 0xe0a0000: native
trace:module:load_dll Loaded module L"C:\\windows\\system32\\xinput1_3.dll" (native) at 0xe0a0000
trace:module:process_attach (L"xinput1_3.dll",(nil)) - START
trace:module:MODULE_InitDLL (0xe0a0000 L"xinput1_3.dll",PROCESS_ATTACH,(nil)) - CALL
```
The native xinput_3.dll is loaded but not due to the override but just because the builtin tells wine that if the native is present it should be loaded instead
```
trace:module:LdrGetDllHandle L"kernel32.dll" -> 0x7b810000 (load path L"C:\\Program Files (x86)\\Steam;.;C:\\windows\\system32;C:\\windows\\system;C:\\windows;C:\\windows\\system32;C:\\windows;C:\\windows\\system32\\wbem")
trace:module:LdrGetDllHandle L"kernel32.dll" -> 0x7b810000 (load path L"C:\\Program Files (x86)\\Steam;.;C:\\windows\\system32;C:\\windows\\system;C:\\windows;C:\\windows\\system32;C:\\windows;C:\\windows\\system32\\wbem")
trace:module:GetModuleFileNameW L"C:\\Program Files (x86)\\Steam\\Steam.exe"
trace:module:MODULE_InitDLL (0xe0a0000,PROCESS_ATTACH,(nil)) - RETURN 1
trace:module:process_attach (L"xinput1_3.dll",(nil)) - END 
```

**Comment by slackner** [comment]
@titan73: An empty string should disable the corresponding dll, which also works fine here locally.

**Comment by slackner** [comment]
@titan73: The embedded "\0"s are actually not necessary, this is a bug in regedit (https://bugs.winehq.org/show_bug.cgi?id=37575). Nevertheless it shouldn't matter in this case.

**Comment by karolherbst** [comment]
@titan73
"tries to load the native but the native dll does not exist" fun fact: it does. DmC required the native DLL and the game worked before and was loading native .NET before I was toying around with GTA IV

Now I get this issue (never tested it before):
```
Error parsing C:\windows\Microsoft.NET\Framework\v4.0.30319\config\machine.config
Parser returned error 0x80004001.
```
But the game still works.

**Comment by karolherbst** [comment]
The effects of the bug also happens with unpatched wine. Allthough I can't say if GTA IV will still brick my prefix.

**Comment by titan73** [comment]
Ah yeah indeed this code is really a mess!

trace:module:load_native_dll Trying native dll L"C:\\windows\\system32\\mscoree.dll"
trace:module:map_image mapped PE file at 0x79000000-0x7904a000
trace:module:map_image mapping section .text at 0x79001000 off 400 size 40000 virt 3ff8d flags 60000020
trace:module:map_image mapping section .data at 0x79041000 off 40400 size 3200 virt 3a78 flags c0000040
trace:module:map_image clearing 0x79044200 - 0x79045000
trace:module:map_image mapping section .rsrc at 0x79045000 off 43600 size 800 virt 780 flags 40000040
trace:module:map_image clearing 0x79045800 - 0x79046000
trace:module:map_image mapping section .reloc at 0x79046000 off 43e00 size 3600 virt 34dc flags 42000040
trace:module:map_image clearing 0x79049600 - 0x7904a000

**Comment by karolherbst** [comment]
:D obviously...

I will try to reproduce it on a clean prefix and will document everything I do in order.

**Comment by titan73** [comment]
To make sure the right user.reg is read "strace -f wine ..." ca be used.

**Comment by slackner** [comment]
@karolherbst: Well, I somehow doubt that it is directly related to changing settings in GTA. What might be possible though is that GTA (or at least a background thread) crashes and that somehow corrupts the registry.

**Comment by titan73** [comment]
Btw how many years is it needed to fix such a trivial bug in regedit. Can we have a fix for that in wine-staging?

**Comment by slackner** [comment]
@titan73: we can have a fix for the regedit issue of course, we just have to find a good solution. Seems like there was already an attempt to fix it: https://www.winehq.org/pipermail/wine-patches/2014-November/135717.html - nevertheless we should probably keep this other issue separate from this bug report. ;)

**Comment by karolherbst** [comment]
@slackner you are right, but the issue is there nethertheless: I tried to be as minimal as possible:

```
1. winetricks steam
2. let steam update
3. quit steam before login
4. move backuped steamapps over
5. start winecfg: enable "emulate virtual desktop" (I don't want wine to mess with my resolution, GTA IV will start in a really small one)
6. start steam and login
7. start GTA IV (which installs all the libraries it needs first through steam)
8. quit GTA IV and steam
9. corrupted prefix!!!
```

**Comment by karolherbst** [comment]
Maybe it is caused by some of the things steam installs before starting GTA IV. I oversaw something with secureROM?

**Comment by titan73** [comment]
Is the dll override for dwrite the same before and after?

**Comment by karolherbst** [comment]
winetricks steam adds "*dwrite=disabled" automatically

**Comment by karolherbst** [comment]
but yes, it is the same

**Comment by karolherbst** [comment]
so I have a trace file now and "grep user.reg trace" gives me this:
```
[pid  3485] open("user.reg", O_RDONLY)  = 18
[pid  3485] open("user.reg", O_WRONLY)  = 267
[pid  3485] lstat("user.reg", {st_mode=S_IFREG|0644, st_size=75621, ...}) = 0
[pid  3485] rename("regd9d0000.tmp", "user.reg") = 0
[pid  3485] open("user.reg", O_WRONLY)  = 400
[pid  3485] lstat("user.reg", {st_mode=S_IFREG|0644, st_size=75621, ...}) = 0
[pid  3485] rename("regd9d0000.tmp", "user.reg") = 0
open("user.reg", O_WRONLY)              = 8
lstat("user.reg", {st_mode=S_IFREG|0644, st_size=75621, ...}) = 0
rename("regd9d0000.tmp", "user.reg")    = 0
```

anything else I should try? The file is 750MB big, likely because of starting steam and dwrite.

**Comment by titan73** [comment]
What is the chdir("xxx") lines before open("user.reg", O_RDONLY) ?

**Comment by karolherbst** [comment]
for the first one: 
```
6310:[pid  3464] chdir("/home/karol/.local/share/wineprefixes/steam") = 0
6312:[pid  3464] chdir("/tmp/.wine-1000/server-802-25cc089") = 0
7104:[pid  3484] chdir("/home/karol/.local/share/wineprefixes/steam") = 0
7110:[pid  3484] chdir("/tmp/.wine-1000/server-802-25cc089") = 0
7159:[pid  3485] fchdir(4)                   = 0
7177:[pid  3464] fchdir(3)                   = 0
7763:[pid  3485] open("user.reg", O_RDONLY)  = 18
```

but I could check the others if you want.

**Comment by slackner** [comment]
could you maybe compress the log, and send them to me/us? feel free to censor any confidential information, and to send them to us privately instead of attaching at the bugtracker. Moreover, it probably also helps to see the full content of the user.reg file.

**Comment by karolherbst** [comment]
trace: https://karolherbst.de/owncloud/public.php?service=files&t=afb3f795255fc5c101de507e4fcf7a3a
user.reg: https://karolherbst.de/owncloud/public.php?service=files&t=0de0fc5c6b2d70291e5379b4a3088ec3

I couldn't find anything important and these files, there are pretty plain, because of clean prefix (the user.reg). The domain is using a self-signed certificate, so you can simply ignore the SSL error there.

**Comment by slackner** [comment]
Confirming, marking as upstream bug. I've found some nice ways to reproduce it and will take a closer look.

**Comment by karolherbst** [comment]
so there are other ways to trigger it?

**Comment by karolherbst** [comment]
Would be nice to know what this bug is causing, because I don't really looking forward to recreate my prefix, just to play some GTA

**Comment by slackner** [comment]
The problem seems to be related to wow64 prefixes. This is probably also why noone of has triggered it yet, we're all using pure 32-bit prefixes most of the time (because 64-bit support in general is still a bit unstable). I haven't figured out yet how exactly to trigger it reliable, but I can reproduce your issue right now, and will try to figure out what causes it.

Since developing 64-bit wine is always a bit complicated (you first need to compile the 64-bit part, then the 32-bit part, then install the 32-bit part, and afterwards the 64-bit part) I do not even have my build environment set up for that. Currently recompiling wine as 64-bit, then I'll add some additional traces to figure out why it fails to find those registry keys.

**Comment by karolherbst** [comment]
ohh, I gave up using win32 prefixes, because winetricks seems to mess with me while trying it:
```
------------------------------------------------------
wine cmd.exe /c echo '%ProgramFiles%' returned empty string
------------------------------------------------------
```
and I never found a way around this issue, so I just used a win64 prefix the entire time, without any bad results until now.

**Comment by slackner** [comment]
@karolherbst: Could you please try out this patch (on top of vanilla wine): http://ix.io/fJE
In order to be included as a wine-staging patch it might require some rebasing.

**Comment by slackner** [comment]
I looked a bit more through the source, and even though my patch seems to solve the problem, it is probably a bit more complicated. When starting with a fresh prefix, then 32-bit and 64-bit overrides are still identical, and changing entries in winecfg is sufficient. After the prefix corruption (how to trigger it?) the existence of a "Wow6432Node" hides the whole original 64-bit registry dll overrides key. This major flaw could also explain a lot of other issues related to 64-bit wineprefixes because it is probably not limited to dll overrides.

As long as this issue is not fixed yet I would probably recommend to use a pure 32-bit wineprefix instead.

**Comment by karolherbst** [comment]
I think one of the installers from GTA might trigger it. I could try out which of them is doing this.

**Comment by karolherbst** [comment]
@slackner yeah, the patch helps, thanks

**Comment by deri82** [comment]
This could be related 

http://source.winehq.org/patches/data/111094

"From: Andrew Eikum <aeikum@codeweavers.com>
Subject: [PATCH] mmdevapi: Always use 64-bit registry keys
Message-Id: <20150501181112.GW20764@foghorn.codeweavers.com>
Date: Fri, 1 May 2015 13:11:12 -0500

winecfg uses mmdevapi's propertystore to save changes to each device's
channel configuration. Since wine64's winecfg is 64-bit, this had the
problem of only updating the 64-bit configuration, and leaving the
32-bit at the default speaker layout.  This manifested as Bugs 38313 and
38497.

After this patch, we always use the 64-bit keys and avoid this problem.

This also matches Windows. On 64-bit Windows, these values are always
stored in the 64-bit registry."

**Comment by slackner** [comment]
Hi everyone,

About one year ago, we announced that we would like to move to more advanced bug tracker. We are sorry for the inconvenience, but now its time to say finally goodbye to the old one. This means, all remaining opened bugs (including this one!) will be closed, and no longer available. If you still need any information from the old bugtracker, please write a mail to webmaster@fds-team.de.

More important, if your issue still exists, please open a new bug report at one of the places mentioned below. For bug reports, please attach terminal output from a recent Wine version (preferably >= 1.8). Please note that in the meantime, Wine Staging has become an official part of WineHQ, so if your issue is already tracked at winehq.org, there is NO NEED to open a new bug report for it. Feel free to add the Wine Staging maintainers (sebastian@fds-team.de, michael@fds-team.de) as CC if you want us to take a look at your bug report.

**NEW BUG TRACKERS:**

* For all Wine AND Wine Staging bugs. USE THIS IF IN DOUBT. (*)
  https://bugs.winehq.org/enter_bug.cgi?product=Wine

* For packaging issues:
  https://bugs.winehq.org/enter_bug.cgi?product=Packaging

* For patch submission:
  https://bugs.wine-staging.com/enter_bug.cgi

Best regards,
Sebastian

----

(*) The "Wine Staging" product in the winehq bugtracker should only be used, when your problem turns out to be a regression introduced in "Wine Staging". Since most problems also occur with Wine, it is recommended to open your bug in the "Wine" product, unless someone can show that it doesn't occur with the development branch.
